### PR TITLE
Fixed #27676 -- Allowed BLOB/TEXT defaults on MariaDB 10.2.1+.

### PR DESCRIPTION
**Pull request by adamchainz**
[Ticket](https://code.djangoproject.com/ticket/27676)

### CODE REVIEW

**Comment by adamchainz** [comment]
I guess this doesn't need a release note since the master branch is 3.0 and has the "now supports MariaDB" release note.

File: django/contrib/gis/db/backends/mysql/schema.py [hunk 1]
```diff
@@ -16,11 +16,15 @@ def __init__(self, *args, **kwargs):
...
 
     def skip_default(self, field):
-        return (
-            super().skip_default(field) or
-            # Geometry fields are stored as BLOB/TEXT and can't have defaults.
-            isinstance(field, GeometryField)
-        )
+        # Geometry fields are stored as BLOB/TEXT
+        # MySQL and MariaDB < 10.2.1 don't support defaults for BLOB or TEXT
+        # https://mariadb.com/kb/en/library/blob-and-text-data-types/
```
Comment (felixxm) [inline django/contrib/gis/db/backends/mysql/schema.py]: Chop this link.

File: django/db/backends/mysql/schema.py [hunk 2]
```diff
@@ -48,7 +48,11 @@ def _is_limited_data_type(self, field):
         return db_type is not None and db_type.lower() in self.connection._limited_data_types
 
     def skip_default(self, field):
-        return self._is_limited_data_type(field)
+        # MySQL and MariaDB < 10.2.1 don't support defaults for BLOB or TEXT
+        # https://mariadb.com/kb/en/library/blob-and-text-data-types/
```
Comment (felixxm) [inline django/db/backends/mysql/schema.py]: Chop this link.

File: django/db/backends/mysql/schema.py [hunk 3]
```diff
@@ -48,7 +48,11 @@ def _is_limited_data_type(self, field):
...
+        if not self.connection.mysql_is_mariadb or self.connection.mysql_version < (10, 2, 1):
+            return self._is_limited_data_type(field)
+        return False
```
Comment (felixxm) [inline django/db/backends/mysql/schema.py]: I would revert this, i.e.:
```python
if self.connection.mysql_is_mariadb and self.connection.mysql_version >= (10, 2, 1):
    return True
return self._is_limited_data_type(field)
```

File: django/db/backends/mysql/schema.py [hunk 4]
```diff
@@ -48,7 +48,11 @@ def _is_limited_data_type(self, field):
         return db_type is not None and db_type.lower() in self.connection._limited_data_types
 
     def skip_default(self, field):
-        return self._is_limited_data_type(field)
+        # MySQL and MariaDB < 10.2.1 don't support defaults for BLOB or TEXT
+        # https://mariadb.com/kb/en/library/blob-and-text-data-types/
+        if not self.connection.mysql_is_mariadb or self.connection.mysql_version < (10, 2, 1):
```
Comment (felixxm) [inline django/db/backends/mysql/schema.py]: Maybe we could add a cached property with this check in `schema.py`, e.g.
```python
@cached_property
def _support_default_on_limited_data_type(self):
    return self.connection.mysql_is_mariadb and self.connection.mysql_version > (10, 2)
```

**Comment by felixxm** [comment]
@adamchainz Yes, release notes are not necessary.

File: django/db/backends/mysql/schema.py (hunk 4 above)
Comment (adamchainz) [reply django/db/backends/mysql/schema.py]: I don't think `cached_property` is worth it, it's a really fast few binary comparisons - the underlying properties are cached.

**Comment by adamchainz** [comment]
Why remove the tests? They are really fast

File: django/db/backends/mysql/schema.py (hunk 3 above)
Comment (adamchainz) [reply django/db/backends/mysql/schema.py]: I think it's better as-is, that we should consider the happy path to be that defaults don't need skipping, and older versions of MySQL/MariaDB need defaults skipping - especially after https://code.djangoproject.com/ticket/30712

**Comment by felixxm** [comment]
I don't see much value in these tests (I know that they're pretty and sophisticated, `ExitStack` etc. really nice :+1: , but still...), we have e.g. many db features which depend on db versions but we don't test them with mock. Moreover CI uses MySQL 5.7 and MariaDB 10.2.24 so both paths are already tested.

File: django/db/backends/mysql/schema.py (hunk 4 above)
Comment (felixxm) [reply django/db/backends/mysql/schema.py]: So maybe without cache, with a separate property we will be able to fix in only one place after [30712](https://code.djangoproject.com/ticket/30712).

**Comment by charettes** [comment]
I agree with @felixxm, the tests are sophisticated and surely pass but they diverge from how we've tested these changes in the past. Not saying that we should not test dynamic feature flags more appropriately but I don't see a reason to start doing it for this particular feature.

File: django/db/backends/mysql/schema.py (hunk 4 above)
Comment (adamchainz) [reply django/db/backends/mysql/schema.py]: Done

**Comment by adamchainz** [comment]
I removed the tests. Yes they were testing the `if` and `or` but tbf I got them wrong the first time, so they at least helped then... :)

**Comment by felixxm** [comment]
@adamchainz Thanks :+1:

**Pull request merged**